bound the maximum number of validators considered for withdrawals per sweep (#3095)
ethereum/consensus-specs: specs/capella/beacon-chain.md

| - | - | - |
| `MAX_WITHDRAWALS_PER_PAYLOAD` | `uint64(2**4)` (= 16) | Maximum amount of withdrawals allowed in each payload |

### Withdrawals processing

| Name | Value |
| - | - |

> | `MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP` | `1024` (= 2**10 ) |

Reviewer: ```suggestion
| `MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP` | `16384` (= 2**14 ) |
```

    for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
        assert withdrawal == expected_withdrawal
        decrease_balance(state, withdrawal.validator_index, withdrawal.amount)

>     if len(expected_withdrawals) == 0:

Reviewer 2: ```suggestion
    if len(expected_withdrawals) < MAX_WITHRAWALS_PER_PAYLOAD:
```

Author: if there are more than 0 withdrawals, then the usual process takes over where you just go via the last validator you could get to

if we _always_ skip forward by the sweep bound then we would end up jumping around the validator set and lose the sequential nature of the scan

Reviewer 2: If you didn't achieve the maximum then we know for a fact that no validator was withdrawable this block in that sweep. There's no need to rescan in the next slot just to include the possible one that may have become withrawable during this one. It's more effective to just add here and continue from where we scanned in this loop

Reviewer 2: In other words, given your argument, why not just add 1 then when there are zero, instead of the full max bound? There may be validators that become withrawable in the next slot and adding one already keeps us alive

Author: > If you didn't achieve the maximum then we know for a fact that no validator was withdrawable this block in that sweep.

no?

let's start at validator index 0, sweep 256 with block max of 16. 

if you get 0 withdrawals, then no validator was withdrawable so we advance the state of the sweep by 256
if we get 16 withdrawals, then it seems most fair to just start at validator index 16 next time
if you get some other number of withdrawals, then I think you want to start just after the index of the last withdrawal you got...

the only special case is when you have 0 withdrawals which this PR currently handles

I would not want to just add 1 to the validator index as that seems that we would duplicate inspections of validators per epoch (keep in mind that withdrawable status only changes at epoch boundaries, unless there is a deposit or if you are slashed which I'll argue are infrequent enough to not bias the 'fairness' of this choice of scan significantly)

Reviewer 2: > if you get some other number of withdrawals, then I think you want to start just after the index of the last withdrawal you got...
> 
I disagree with this, if you get less than 16 it means that you scanned 256 and found no withdrawable validators and you are making the point below that it's a good assumption that they will not be withdrawable the next slot, so I want to start next sweep at 256 more, and not the last index. 

> I would not want to just add 1 to the validator index as that seems that we would duplicate inspections of validators per epoch (keep in mind that withdrawable status only changes at epoch boundaries, unless there is a deposit or if you are slashed which I'll argue are infrequent enough to not bias the 'fairness' of this choice of scan significantly)

Author: ok I see your point and I'll update it, thanks for explaining!

    for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
        assert withdrawal == expected_withdrawal
        decrease_balance(state, withdrawal.validator_index, withdrawal.amount)

    # Update the next withdrawal index if this block contained withdrawals
    if len(expected_withdrawals) != 0:
        latest_withdrawal = expected_withdrawals[-1]
        state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)

    # Update the next validator index to start the next withdrawal sweep

>     if len(expected_withdrawals) == MAX_WITHDRAWALS_PER_PAYLOAD:

Reviewer 3: For the `len(expected_withdrawals) == MAX_WITHDRAWALS_PER_PAYLOAD` case `state.next_withdrawal_index` is updated twice. Is that okay? Why not do
```
if len == MAX_WITHDRAWALS_PER_PAYLOAD
elif len > 0
else
```

Reviewer 2: Where is the second update? I only see it set once.

Reviewer 3: Oh, miss-read `next_withdrawal_index` for `next_withdrawal_validator_index`. All good then :+1:

Reviewer: the names are confusing to read as is :/

Author: agreed :/

        # Next sweep starts after the latest withdrawal's validator index

>         latest_withdrawal = expected_withdrawals[-1]
>         next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))

Reviewer: suggest the following for consistency (it is 117 chars so in linter limit)
```suggestion
        next_validator_index = ValidatorIndex((expected_withdrawals[-1].validator_index + 1) % len(state.validators))
```

    for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
        assert withdrawal == expected_withdrawal
        decrease_balance(state, withdrawal.validator_index, withdrawal.amount)

    if len(expected_withdrawals) != 0:
        latest_withdrawal = expected_withdrawals[-1]
        state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)

    if len(expected_withdrawals) < MAX_WITHDRAWALS_PER_PAYLOAD:
        # advance sweep if there were not a full set of withdrawals
        next_index = state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP
        next_validator_index = next_index % len(state.validators)
        state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
    else:

>         latest_withdrawal = expected_withdrawals[-1]

Reviewer 2: ```suggestion
```

Reviewer: You need the assignment in there or an outer assignment beacuse the prior if isn't necessarily hit. I cover that in my comment below

Reviewer 2: > You need the assignment in there or an outer assignment beacuse the prior if isn't necessarily hit. I cover that in my comment below

no because this else is only hit when `len(expected_withdrawals) >= MAX_WITHDRAWALS_PER_PAYLAOD`  in which case the previous assignment already was hit.

Author: so I'm trying to avoid python's dynamic scoping here and so just elected to make the assignment twice

if we are ok w/ imo non-standard (relative to what most readers are assumed to know) scoping behavior then we can drop the extra assignment

        next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))

>         state.next_withdrawal_validator_index = next_validator_index

Author: I can add more comments but we can't pull out the `latest_withdrawal = expected_withdrawals[-1]` assignment from its current place bc `len(expected_withdrawals) == 0` is possible which would raise an error

and there is no 'safe' built-in method for indexing on lists so there isn't some kind of pythonic thing we could do (which imo we don't want to actually do)

>         state.next_withdrawal_validator_index = next_validator_index

Reviewer: hmm it isn't safe as currently written either though. The `else` isn't scoped by non-zero expected_withdrawals. will write an updated suggestion

>         state.next_withdrawal_validator_index = next_validator_index

Reviewer 2: > hmm it isn't safe as currently written either though. The `else` isn't scoped by non-zero expected_withdrawals. will write an updated suggestion

It is, I'm fine with @ralexstokes' intent to explicitly assign in case people get confused by the scoping, but it's perfectly valid to not have the second `latest_withdrawal` assignment.
```shell
$cat danny.py
danny = 1
if 1 > 0:
    danny = 4
if 2 > 3:
    danny = 1
else:
    print(danny)

$ python danny.py
4
```
```

> 
>     if len(expected_withdrawals) != 0:
>         latest_withdrawal = expected_withdrawals[-1]
>         state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
> 
>     if len(expected_withdrawals) < MAX_WITHDRAWALS_PER_PAYLOAD:
>         # advance sweep if there were not a full set of withdrawals
>         next_index = state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP
>         next_validator_index = next_index % len(state.validators)
>         state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
>     else:
>         latest_withdrawal = expected_withdrawals[-1]
>         next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))
>         state.next_withdrawal_validator_index = next_validator_index

Reviewer: okay, how about this. I really don't like relying on python's scoping rules. If not this, then I'd do two var assignments in different places 

```suggestion
    
    # Update the next withdrawal index if this block contained withdrawals
    if len(expected_withdrawals) != 0:
        state.next_withdrawal_index = WithdrawalIndex(expected_withdrawals[-1].index + 1)

    # Update the next validator index to start the next withdrawal sweep
    if len(expected_withdrawals) == MAX_WITHDRAWALS_PER_PAYLOAD:
        # Next sweep starts after the latest withdrawal's validator index
        next_validator_index = ValidatorIndex((expected_withdrawals[-1].validator_index + 1) % len(state.validators))
        state.next_withdrawal_validator_index = next_validator_index
    else:
        # Advance sweep by the max length of the sweep if there was not a full set of withdrawals
        next_index = state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP
        next_validator_index = next_index % len(state.validators)
        state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
```

>         state.next_withdrawal_validator_index = next_validator_index

Author: this seems fine, I've made the update

    for expected_withdrawal, withdrawal in zip(expected_withdrawals, payload.withdrawals):
        assert withdrawal == expected_withdrawal
        decrease_balance(state, withdrawal.validator_index, withdrawal.amount)

>     if len(expected_withdrawals) == 0:

Reviewer 3: Should advance the pointer every time withdrawals are not maxed
```suggestion
    if len(expected_withdrawals) < MAX_WITHDRAWALS_PER_PAYLOAD:
```

        # advance sweep if there were no withdrawals
        next_validator_index = (state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP) % len(state.validators)
        state.next_withdrawal_validator_index = ValidatorIndex(next_validator_index)
        return

    latest_withdrawal = expected_withdrawals[-1]
    state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
    next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))

>     state.next_withdrawal_validator_index = next_validator_index

Reviewer 3: Should be changed to always move the pointer forward. For the case of 0 withdrawals, incomplete withdrawals, and max withdrawals.

```py
    if len(expected_withdrawals) < MAX_WITHDRAWALS_PER_PAYLOAD:
        next_validator_index = (state.next_withdrawal_validator_index + MAX_VALIDATORS_PER_WITHDRAWALS_SWEEP) % len(state.validators)
    else:
        next_validator_index = ValidatorIndex((expected_withdrawals[-1].validator_index + 1) % len(state.validators))

    state.next_withdrawal_validator_index = next_validator_index
    state.next_withdrawal_index = WithdrawalIndex(state.next_withdrawal_index + len(expected_withdrawals))
```

Reviewer 3: I don't understand why the full sweep can only happen in the 0 withdrawal case and not in the incomplete withdrawals case. You be scanning a range of validators twice in consecutive blocks due to not updating the pointer position.

        state.next_withdrawal_index = WithdrawalIndex(latest_withdrawal.index + 1)
        next_validator_index = ValidatorIndex((latest_withdrawal.validator_index + 1) % len(state.validators))
        state.next_withdrawal_validator_index = next_validator_index

>     else:

Reviewer: I kind of prefer this is a blocking condition (as it is the exception)

e.g.
```python
if len(expected_withdrawals == 0):
    next_val_index = ...
    state.next_val_index = ...
    return

latest_withdrawal = expected_ ...
state.next_withdrawal =...
